### ISSUE TITLE
questions for using multiple outputs

### ISSUE THREAD

**Issue by liuxindazz** [pytorch/ignite#353]
Question 1:
In pytorch, my model is just 
`output1, output2 = model(inputs)`
`loss = criterion['1'](output1, labels)+criterion['2'](output2, labels)`

How should I do if I want to use ignite?
I wrote the code like this:
`criterion={'1':loss1, '2':loss2}`
`trainer = create_supervised_trainer(model, optimizer, criterion, device=device)`
but got the type error:` 'dict' object is not callable.`

Question 2:
I don't understand the usage of `output_transform`.
I don't know the output format when I use `create_supervised_trainer` and `create_supervised_trainer`,  so how to define the output_transform function?
Could someone give me an example?

**Comment by vfdev-5** [comment]
@liuxindazz 

> Question 1:

There are two ways to do this:

#### Using `create_supervised_trainer`
Criterion should be callable, so you can implement a callable that computes the loss as you described:

```python
model = ...
optimizer = ...
loss1 = ...
loss2 = ...

def criterion(outputs, labels):
      output1, output2 = outputs
     loss = loss1(output1, labels) + loss2(output2, labels)
     return loss.item()
 
trainer = create_supervised_trainer(model, optimizer, criterion, device=device)
```

#### Without `create_supervised_trainer`

In this case you have more flexibilty and you can better control model update step. So you need to implement your own processing function (see details [here](https://pytorch.org/ignite/concepts.html#engine)) like this:
```python

model = ...
criterion={'1':loss1, '2':loss2}
optimizer = ...

def update_model(trainer, batch):
     inputs, labels = batch
     optimizer.zero_grad()
     output1, output2 = model(inputs)
     loss = criterion['1'](output1, labels)+criterion['2'](output2, labels)
     loss.backward()
     optimizer.step()
     return loss.item()

trainer = Engine(update_model)
```

> Question 2:

In the previous example the output of `update_model` is `loss.item()` and it is written in `trainer.state.output`. For `create_supervised_trainer` the output is also just `loss.item()`, written to `trainer.state.output`.

Actually, you can add whatever you want when you return from `update_model`. For example,
```python
def update_model(trainer, batch):
     inputs, labels = batch
     ...
     return {'loss': loss.item(), 
                 'output1': output1, 
                 'output2': output2, 
                 'inputs': inputs, 
                 'labels': labels}
```
so `trainer.state.output` after each iteration will contain `loss.item(), output1, output2, inputs, labels`
When you would like to compute for example running average metrics on this you need to specify
`output_transform`:
```python
output_transform1 = lambda data: data['output1'], data['labels']
output_transform2 = lambda data: data['output2'], data['labels']

alpha = 0.98
acc_metric1 = RunningAverage(Accuracy(output_transform=output_transform1), alpha=alpha)
acc_metric1.attach(trainer, 'running_avg_accuracy1')

acc_metric2 = RunningAverage(Accuracy(output_transform=output_transform2), alpha=alpha)
acc_metric2.attach(trainer, 'running_avg_accuracy2')

@trainer.on(Events.ITERATION_COMPLETED)
def log_running_avg_metrics(engine):
    print("running avg accuracy:", engine.state.metrics['running_avg_accuracy1'])
    print("running avg accuracy:", engine.state.metrics['running_avg_accuracy2])
```

More on this take a look [here](https://pytorch.org/ignite/metrics.html) and [here](https://pytorch.org/ignite/metrics.html#ignite.metrics.RunningAverage).

HTH

**Comment by liuxindazz** [comment]
@vfdev-5 Thanks, it's a great help to me.
